functionary/prompt.py

```diff
@@ -31,147 +187,543 @@ def from_message(cls, message: Dict) -> EndToken:
-    Args:
-        messages (List[Dict]): list of messages where each message is in the format of OpenAI
-        functions (Optional[List[Dict]]): list of functions where each function is in the format of OpenAI
+def get_default_prompt_template() -> PromptTemplate:
+    """Return default prompt template to be used
 
     Returns:
-        str: the final prompt that will be used.
+        _type_: _description_
     """
-    messages_clone = messages.copy()  # To avoid modifying the original list
-
-    if functions is None:
-        functions = []
-
-    messages_clone.insert(0, {"role": "system", "content": generate_schema_from_functions(functions)})
-    messages_clone.insert(1, {"role": "system", "content": SYSTEM_MESSAGE})
-
-    full_text = ""
-    for message in messages_clone:
-        full_text += get_text_from_message(message)
-    return full_text.strip()
+    return PromptTemplateV2.get_prompt_template()
 
 
-def get_end_token_to_token_id(tokenizer: Any) -> Dict[EndToken, int]:
-    """return a dictionary mapping from end_token --> token_id
-
-    Args:
-        tokenizer (Any): tokenizer in transformers
-
-    Returns:
-        Dict[int, EndToken]: the mapping from token_id --> end_token
-    """
-    result = {}
-    for item in EndToken:
-        tok_ids = tokenizer.encode(item.value, add_special_tokens=False)
-        assert len(tok_ids) <= 2
-        if len(tok_ids) == 2:
-            assert tok_ids[0] == 29871  # Llama tokenizer adds this token intentionally
-        result[item] = tok_ids[-1]
-    return result
+def get_prompt_template(version: int) -> PromptTemplate:
+    if version == "1":
+        return PromptTemplateV1.get_prompt_template()
+    return PromptTemplateV2.get_prompt_template()
 
 
-def get_number_of_tokens_of_prefix_assistant(tokenizer: Any) -> int:
-    """This function is used to compute the number of tokens of "\nassistant" in full prompt
+def get_prompt_template_from_tokenizer(tokenizer: Any) -> PromptTemplate:
+    """This function will determine the prompt template based on tokenizer.
+    Under the hood, this function will check if tokenizer contains some special tokens from template or not
 
     Args:
-        tokenizer (_type_): The tokenizer used for tokenizing
+        tokenizer (Any): Tokenizer
 
     Returns:
-        _type_: number of tokens of "\nassistant" in full prompt
+        _type_: _description_
     """
-    text1 = "<|END_OF_USER|>\nassistant"
-    tok_ids1 = tokenizer.encode(text1, add_special_tokens=False)
-    text2 = "<|END_OF_USER|>"
-    tok_ids2 = tokenizer.encode(text2, add_special_tokens=False)
-    return len(tok_ids1) - len(tok_ids2)
+    p1 = PromptTemplateV1.get_prompt_template()
+    p2 = PromptTemplateV2.get_prompt_template()
+    token_ids = tokenizer.encode(p1.start_function, add_special_tokens=False)
+    if token_ids[0] == 29871:
+        token_ids = token_ids[1:]
+    if len(token_ids) == 1:
+        return p1
+    return p2
```

Comment:
Do you think its easier and more scalable to get the jinja chat_template from the tokenizer itself and find the presence of certain special tokens from there instead?

Reply:
yes we can but need to make sure that chat_template is saved in tokenizer at the training, we can change this later once all tokenizers of models already contain chat_template

Reply:
Ok